File: epi_judge_python/power_x_y.py
from test_framework import generic_test


def power(x: float, y: int) -> float:
    # TODO - you fill in here.
    # break y into binary, sum of 2's powers
    tempx = 0
    result = 1
    yneg = y < 0
    if yneg:
        y = -y
    while y > 0:
        if not tempx:
            if not yneg:
                tempx = x
            else:
                tempx = 1 / x
        else:
            tempx *= tempx
        if y & 1 > 0:
            result *= tempx
        y >>= 1

    return result


if __name__ == "__main__":
    exit(generic_test.generic_test_main("power_x_y.py", "power_x_y.tsv", power))

File: epi_judge_cpp/power_x_y.cc
#include "test_framework/generic_test.h"
double Power(double x, int y) {
  // TODO - you fill in here.
  // return 0.0;

  double result{1.0};
  bool negative{false};

  if (y < 0) {
    negative = true;
    y = -y;
  }
  while (y) {
    // std::cout << "y"  << y << std::endl;
    if (y & 1) {
      result *= x;
    }
    y >>= 1;
    x *= x;
  }

  return negative ? 1 / result : result;
}

int main(int argc, char *argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"x", "y"};
  return GenericTestMain(args, "power_x_y.cc", "power_x_y.tsv", &Power,
                         DefaultComparator{}, param_names);
}
